ergotree-ir/src/mir/constant.rs

```diff
@@ -142,6 +144,76 @@ impl<T: LiftIntoSType + Into<Literal>> From<Option<T>> for Literal {
         Literal::Opt(Box::new(opt.map(|e| e.into())))
     }
 }
+
+impl TryFrom<Value> for Constant {
+    type Error = String;
+    #[allow(clippy::unwrap_used)]
+    fn try_from(value: Value) -> Result<Self, Self::Error> {
+        match value {
+            Value::Boolean(b) => Ok(Constant::from(b)),
+            Value::Byte(b) => Ok(Constant::from(b)),
+            Value::Short(s) => Ok(Constant::from(s)),
+            Value::Int(i) => Ok(Constant::from(i)),
+            Value::Long(l) => Ok(Constant::from(l)),
+            Value::BigInt(b) => Ok(Constant::from(b)),
+            Value::SigmaProp(s) => Ok(Constant::from(*s)),
+            Value::GroupElement(e) => Ok(Constant::from(*e)),
+            Value::CBox(i) => Ok(Constant::from(i)),
+            Value::Coll(coll) => {
+                let (v, tpe) = match coll {
+                    CollKind::NativeColl(n) => (
+                        Literal::Coll(CollKind::NativeColl(n)),
+                        SType::SColl(Box::new(SType::SByte)),
+                    ),
+                    CollKind::WrappedColl { elem_tpe, items } => {
+                        let mut new_items = Vec::with_capacity(items.len());
+                        for v in items {
+                            let c = Constant::try_from(v)?;
+                            new_items.push(c.v);
+                        }
+                        (
+                            Literal::Coll(CollKind::WrappedColl {
+                                elem_tpe: elem_tpe.clone(),
+                                items: new_items,
+                            }),
+                            SType::SColl(Box::new(elem_tpe)),
+                        )
+                    }
+                };
+                Ok(Constant { v, tpe })
+            }
+            Value::Opt(lit) => match *lit {
+                Some(v) => {
+                    let c = Constant::try_from(v)?;
+                    Ok(Constant {
+                        v: Literal::Opt(Box::new(Some(c.v))),
+                        tpe: c.tpe,
+                    })
+                }
+                None => Err("Can't convert from Value::Opt(None) to Constant".into()),
+            },
+            Value::Tup(t) => {
+                if let Ok(t) = t.try_mapped::<_, _, String>(|v| {
+                    let c = Constant::try_from(v)?;
+                    Ok((c.v, c.tpe))
+                }) {
+                    let tuple_items = t.mapped_ref(|(l, _)| l.clone());
+                    let tuple_item_types = SType::STuple(STuple {
+                        items: t.mapped(|(_, tpe)| tpe),
+                    });
+                    Ok(Constant {
+                        v: Literal::Tup(tuple_items),
+                        tpe: tuple_item_types,
+                    })
+                } else {
+                    Err("Can't convert Value:Tup element".into())
+                }
+            }
+            v => Err(format!("Cannot convert {:?} into Constant", v)),
```

Comment:
I suggest not to use a catch-all case but explicitly handle all the cases. This way if enum changes the compiler will force you to handle it.

Reply:
Done.